### ISSUE TITLE
High resource usage during train_policy.py on Colab Pro

### ISSUE THREAD

**Issue by ParisaO** [pietro-talli/tmlcn_code#2]
Hi,

Thanks for sharing this project — it's been very helpful so far!

I'm currently running train_policy.py on Google Colab Pro, but I'm noticing that the training process is extremely resource-intensive. Even when I reduce the number of episodes to 2,000 (from the default 20,000) and set the number of codewords to a single one (not all of them), it still consumes a lot of time and GPU.

Is this expected behavior for this setup, or might there be ways to optimize the training (e.g., batch size, episode length, or architecture)?

Any advice would be appreciated!

**Comment by ParisaO** [comment]
Also, I am plotting the runs_policy train/loss figures and I don't see a decreasing pattern. Is it normal? 

<img width="1099" alt="Image" src="https://github.com/user-attachments/assets/cf5ebb50-60d7-444d-8085-7b92c852d572" />

**Comment by pietro-talli** [comment]
Hi, thanks for reaching out! I’ll need a bit more information to better assist you.
From what I can tell, setting the number of codewords to 1 isn’t valid. The number of codewords define the possible encodings for a feature—setting this to 1 would mean every feature is encoded identically, which prevents the model from learning a meaningful policy. Also, note that vector quantization shouldn't be the most computationally expensive operation. Typically, the most resource-intensive components are the CNN on the sensor side and the LSTM on the actuator side.

Regarding the policy training loss: this can be a bit tricky. In reinforcement learning, it's common for the training loss (often related to value function fitting) not to decrease monotonically, since the policy it's trying to fit is continually changing. Rather than focusing solely on the training loss, it's generally more informative to monitor the reward signal to evaluate learning progress. That said, you should expect the training loss to stabilize once the policy converges.

Are you training on the CartPole environment? What RL algorithm are you using? 
Let me know if you can share more context—happy to help further!

**Comment by ParisaO** [comment]
Hi, thank you very much for your response. I’m running your repository with a few minor modifications, such as using different package versions to ensure compatibility with my device.

I appreciate your clarification regarding the number of codewords. However, even after making that adjustment, I’m still unsure whether the training of the policies (in train_policy.py) is supposed to require this much time and computational power.

I’ll take a closer look at the reward function and follow up with you soon, but I’m also wondering if I might be missing something. Could you share your insight on how the energy or computational cost of training the policies compares to that of training the encoder or quantizers?
For example, in my case, it takes me 4-5 minutes to train a quantizer with a dataset of size 50k on google Colab with A100 GPU,  and I use more than 20 A100 GPU units to run just one train_policy.py only for 2000 episodes (you used 20k episodes)
